File: src/main/java/com/baulsupp/oksocial/authenticator/Authorisation.java
package com.baulsupp.oksocial.authenticator;

import com.baulsupp.oksocial.credentials.CredentialsStore;
import com.baulsupp.oksocial.credentials.ServiceDefinition;
import com.baulsupp.oksocial.output.OutputHandler;
import com.baulsupp.oksocial.secrets.Secrets;
import com.baulsupp.oksocial.util.UsageException;
import java.io.IOException;
import java.util.List;
import java.util.Optional;
import okhttp3.OkHttpClient;

import static java.util.stream.Collectors.joining;

public class Authorisation {
  private ServiceInterceptor interceptor;
  private CredentialsStore credentialsStore;
  private OkHttpClient client;
  private OutputHandler outputHandler;

  public Authorisation(ServiceInterceptor interceptor, CredentialsStore credentialsStore,
      OkHttpClient client, OutputHandler outputHandler) {
    this.interceptor = interceptor;
    this.credentialsStore = credentialsStore;
    this.client = client;
    this.outputHandler = outputHandler;
  }

  public void authorize(Optional<AuthInterceptor<?>> auth, Optional<String> token,
      List<String> authArguments) throws Exception {
    failIfNoAuthInterceptor(auth.isPresent());

    if (token.isPresent()) {
      storeCredentials(auth.get(), token.get());
    } else {
      authRequest(auth.get(), authArguments);
    }
  }

  private void failIfNoAuthInterceptor(boolean present) {
    if (!present) {
      throw new UsageException(
          "unable to find authenticator. Specify name from " + interceptor.names()
              .stream()
              .collect(joining(", ")));
    }
  }

  private <T> void storeCredentials(AuthInterceptor<T> auth, String token) {
    T credentials = auth.serviceDefinition().parseCredentialsString(token);
    credentialsStore.storeCredentials(credentials, auth.serviceDefinition());
  }

  private <T> void authRequest(AuthInterceptor<T> auth, List<String> authArguments)
      throws Exception {

    T credentials = auth.authorize(client, outputHandler, authArguments);

    credentialsStore.storeCredentials(credentials, auth.serviceDefinition());

    Secrets.instance().saveIfNeeded();

    // TODO validate credentials
  }

  public <T> void renew(Optional<AuthInterceptor<T>> auth) throws IOException {
    failIfNoAuthInterceptor(auth.isPresent());

    ServiceDefinition<T> serviceDefinition = auth.get().serviceDefinition();
    Optional<T> credentials = credentialsStore.readDefaultCredentials(serviceDefinition);

    if (!credentials.isPresent()) {
      throw new UsageException("no existing credentials");
    }

    if (!auth.get().canRenew(credentials.get())) {
      throw new UsageException("credentials not renewable");
    }

    Optional<T> newCredentials = auth.get().renew(client, credentials.get());

    if (newCredentials.isPresent()) {
      credentialsStore.storeCredentials(newCredentials.get(), serviceDefinition);
    } else {
      throw new UsageException("failed to renew");
    }
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/authenticator/Authorisation.kt
package com.baulsupp.oksocial.authenticator

import com.baulsupp.oksocial.credentials.CredentialsStore
import com.baulsupp.oksocial.output.OutputHandler
import com.baulsupp.oksocial.output.util.UsageException
import com.baulsupp.oksocial.secrets.Secrets
import okhttp3.OkHttpClient
import okhttp3.Response

class Authorisation(val interceptor: ServiceInterceptor, val credentialsStore: CredentialsStore,
                    val client: OkHttpClient, val outputHandler: OutputHandler<Response>, val defaulttokenSet: String) {

  suspend fun authorize(auth: AuthInterceptor<*>?, token: String?,
                        authArguments: List<String>, tokenSet: String) {
    if (auth == null) {
      throw UsageException(
        "unable to find authenticator. Specify name from " + interceptor.names().joinToString(", "))
    }

    if (token != null) {
      storeCredentials(auth, token, tokenSet)
    } else {
      authRequest(auth, authArguments, tokenSet)
    }
  }

  private fun <T> storeCredentials(auth: AuthInterceptor<T>, token: String, tokenSet: String) {
    val credentials = auth.serviceDefinition().parseCredentialsString(token)
    credentialsStore.set(auth.serviceDefinition(), tokenSet, credentials)
  }

  suspend fun <T> authRequest(auth: AuthInterceptor<T>, authArguments: List<String>, tokenSet: String) {
    auth.serviceDefinition().accountsLink()?.let { outputHandler.info("Accounts: " + it) }

    val credentials = auth.authorize(client, outputHandler, authArguments)

    credentialsStore.set(auth.serviceDefinition(), tokenSet, credentials)

    Secrets.instance.saveIfNeeded()

    // TODO validate credentials
  }

  suspend fun <T> renew(auth: AuthInterceptor<T>?, tokenSet: String) {
    if (auth == null) {
      throw UsageException(
        "unable to find authenticator. Specify name from " + interceptor.names().joinToString(", "))
    }

    val serviceDefinition = auth.serviceDefinition()

    val credentials = credentialsStore.get(serviceDefinition, tokenSet)
      ?: throw UsageException("no existing credentials")

    if (!auth.canRenew(credentials)) {
      throw UsageException("credentials not renewable")
    }

    val newCredentials = auth.renew(client, credentials) ?: throw UsageException("failed to renew")

    credentialsStore.set(serviceDefinition, tokenSet, newCredentials)
  }

  fun remove(auth: AuthInterceptor<*>?, tokenSet: String) {
    if (auth == null) {
      throw UsageException(
        "unable to find authenticator. Specify name from " + interceptor.names().joinToString(", "))
    }

    credentialsStore.remove(auth.serviceDefinition(), tokenSet)
  }
}
